dp/drivers/i40e.c

```diff
@@ -0,0 +1,1003 @@
+	txq->ring_entries = (struct tx_entry *)((uintptr_t) txq->ring +
+			align_up(sizeof(struct i40e_tx_desc) * nb_desc, I40E_RING_BASE_ALIGN));
+	txq->len = nb_desc;
+
+	ret = mem_lookup_page_machine_addr(page, PGSIZE_2MB, &page_phys);
+	if (ret)
+		goto err;
+	txq->ring_physaddr = page_phys +
+			     align_up(sizeof(struct tx_queue), I40E_RING_BASE_ALIGN);
+
+	/* these two fields are initialized in dev_start using DPDK :
+	txq->reg_idx = dtxq->reg_idx;
+	txq->tdt_reg_addr = (uint32_t *)dtxq->qtx_tail; */
+
+	txq->tx_free_thresh = DEFAULT_TX_FREE_THRESH;
+	txq->nb_tx_desc = nb_desc;
+	txq->tx_rs_thresh = DEFAULT_TX_RS_THRESH;
+
+	txq->etxq.reclaim = i40e_tx_reclaim;
+	txq->etxq.xmit = i40e_tx_xmit;
+	i40_reset_tx_queue(txq);
+	dev->data->tx_queues[queue_idx] = &txq->etxq;
+	/* release the dpdk memory location and all its buffers*/
+	//i40e_dev_tx_queue_release(dtxq);
+	return 0;
+
+err:
+	mem_free_page(page, PGSIZE_2MB);
+	return ret;
+
+}
+
+static void init_filter(struct rte_eth_fdir_filter *filter, struct rte_fdir_filter *in)
+{
+	memset(filter, 0, sizeof(*filter));
+
+	assert(in->iptype == RTE_FDIR_IPTYPE_IPV4);
+	assert(in->l4type == RTE_FDIR_L4TYPE_TCP);
+
+	filter->input.flow_type = RTE_ETH_FLOW_NONFRAG_IPV4_TCP;
+	filter->input.flow.tcp4_flow.ip.src_ip = ntoh32(in->ip_src.ipv4_addr);
+	filter->input.flow.tcp4_flow.ip.dst_ip = ntoh32(in->ip_dst.ipv4_addr);
+	filter->input.flow.tcp4_flow.src_port = hton16(in->port_src);
+	filter->input.flow.tcp4_flow.dst_port = hton16(in->port_dst);
+}
+
+static int fdir_add_perfect_filter(struct ix_rte_eth_dev *dev, struct rte_fdir_filter *fdir_ftr, uint16_t soft_id, uint8_t rx_queue, uint8_t drop)
+{
+	int ret;
+	struct rte_eth_fdir_filter filter;
+
+	init_filter(&filter, fdir_ftr);
+
+	filter.action.behavior = drop ? RTE_ETH_FDIR_REJECT : RTE_ETH_FDIR_ACCEPT;
+	filter.action.report_status = RTE_ETH_FDIR_REPORT_ID;
+	filter.action.rx_queue = rx_queue;
+	filter.soft_id = soft_id;
+
+	spin_lock(&dev->lock);
+	ret = rte_eth_dev_filter_ctrl(dev->port, RTE_ETH_FILTER_FDIR, RTE_ETH_FILTER_ADD, &filter);
+	spin_unlock(&dev->lock);
+	return ret;
+}
+
+static int fdir_remove_perfect_filter(struct ix_rte_eth_dev *dev, struct rte_fdir_filter *fdir_ftr, uint16_t soft_id)
+{
+	int ret;
+	struct rte_eth_fdir_filter filter;
+
+	init_filter(&filter, fdir_ftr);
+
+	spin_lock(&dev->lock);
+	ret = rte_eth_dev_filter_ctrl(dev->port, RTE_ETH_FILTER_FDIR, RTE_ETH_FILTER_DELETE, &filter);
+	spin_unlock(&dev->lock);
+	return ret;
+}
+
+static int rss_hash_conf_get(struct ix_rte_eth_dev *dev, struct ix_rte_eth_rss_conf *ix_reta_conf)
+{
+	int ret;
+	struct rte_eth_rss_conf reta_conf;
+
+	ret = rte_eth_dev_rss_hash_conf_get(dev->port, &reta_conf);
+	if (ret < 0)
+		return ret;
+
+	ix_reta_conf->rss_key = reta_conf.rss_key;
+	ix_reta_conf->rss_hf = reta_conf.rss_hf;
+
+	return ret;
+}
+
+static void mac_addr_add(struct ix_rte_eth_dev *dev, struct eth_addr *mac_addr, uint32_t index, uint32_t vmdq)
```

Comment:
There is code duplication. This function is the same as https://github.com/ix-project/ix/blob/master/dp/drivers/ixgbe.c#L857. Can you please identify the duplicated functions between these two files? Then, please move them to https://github.com/ix-project/ix/blob/master/dp/drivers/common.c. I'd like to see a separate pull request where you refactor ixgbe.c in that way.